src/arch/arm/64/kernel/vspace.c

```diff
@@ -1251,134 +977,66 @@ static inline void invalidateTLBByASIDVA(asid_t asid, vptr_t vaddr)
 #endif
 }
 
-pde_t *pageTableMapped(asid_t asid, vptr_t vaddr, pte_t *pt)
-{
-    findVSpaceForASID_ret_t find_ret;
-    lookupPDSlot_ret_t lu_ret;
-
-    find_ret = findVSpaceForASID(asid);
-    if (find_ret.status != EXCEPTION_NONE) {
-        return NULL;
-    }
-
-    lu_ret = lookupPDSlot(find_ret.vspace_root, vaddr);
-    if (lu_ret.status != EXCEPTION_NONE) {
-        return NULL;
-    }
-
-    if (pde_pde_small_ptr_get_present(lu_ret.pdSlot) &&
-        (pde_pde_small_ptr_get_pt_base_address(lu_ret.pdSlot) == pptr_to_paddr(pt))) {
-        return lu_ret.pdSlot;
-    }
-
-    return NULL;
-}
 
-void unmapPageUpperDirectory(asid_t asid, vptr_t vaddr, pude_t *pud)
+void unmapPageTable(asid_t asid, vptr_t vptr, pte_t *target_pt)
 {
-    pgde_t *pgdSlot;
-
-    pgdSlot = pageUpperDirectoryMapped(asid, vaddr, pud);
-    if (likely(pgdSlot != NULL)) {
-        *pgdSlot = pgde_pgde_invalid_new();
-        cleanByVA_PoU((vptr_t)pgdSlot, pptr_to_paddr(pgdSlot));
-        invalidateTLBByASID(asid);
+    findVSpaceForASID_ret_t find_ret = findVSpaceForASID(asid);
+    if (unlikely(find_ret.status != EXCEPTION_NONE)) {
+        /* nothing to do */
+        return;
     }
-}
-
-void unmapPageDirectory(asid_t asid, vptr_t vaddr, pde_t *pd)
-{
-    pude_t *pudSlot;
-
-    pudSlot = pageDirectoryMapped(asid, vaddr, pd);
-    if (likely(pudSlot != NULL)) {
-        *pudSlot = pude_invalid_new();
+    pte_t *ptSlot = NULL;
+    pte_t *pt = (pte_t *)find_ret.vspace_root;
 
-        cleanByVA_PoU((vptr_t)pudSlot, pptr_to_paddr(pudSlot));
-        invalidateTLBByASID(asid);
+    for (word_t i = 0; i < UPT_LEVELS - 1 && pt != target_pt; i++) {
+        ptSlot = pt + GET_UPT_INDEX(vptr, i);
+        if (unlikely(!pte_pte_table_ptr_get_present(ptSlot))) {
+            /* couldn't find it */
+            return;
+        }
+        pt = paddr_to_pptr(pte_pte_table_ptr_get_pt_base_address(ptSlot));
     }
```

Comment:
We might need this loop in a separate function for verification in the future, but let's leave it here for now and see how far we get.